In [22]:
import pandas as pd
import webbrowser
from fyers_apiv3 import fyersModel
from datetime import datetime, timedelta
import time

In [23]:
client_id="W7LM9VJA41-200"
secret_key="qpKCdzLr6qijQPIR"
redirect_uri="https://tarantularesearch.com/"

In [24]:
def generate_fyers_access_token(
    client_id: str,
    secret_key: str,
    redirect_uri: str,
    state: str = "sample_state"
) -> dict:
    """
    Handles FYERS OAuth flow and returns access token response.

    Returns
    -------
    dict
        Full response containing access_token, refresh_token, expiry, etc.
    """

    # -------------------------------
    # Step 1: Generate auth URL
    # -------------------------------
    session = fyersModel.SessionModel(
        client_id=client_id,
        secret_key=secret_key,
        redirect_uri=redirect_uri,
        response_type="code",
        state=state
    )

    auth_url = session.generate_authcode()
    webbrowser.open(auth_url)

    # -------------------------------
    # Step 2: User pastes redirected URL
    # -------------------------------
    raw_code = input("Paste redirected URL here: ").strip()

    try:
        auth_code = raw_code.split("auth_code=")[1].split("&")[0]
    except IndexError:
        raise ValueError("Invalid redirect URL. Auth code not found.")

    # -------------------------------
    # Step 3: Exchange auth_code for access_token
    # -------------------------------
    session = fyersModel.SessionModel(
        client_id=client_id,
        secret_key=secret_key,
        redirect_uri=redirect_uri,
        response_type="code",
        grant_type="authorization_code"
    )

    session.set_token(auth_code)
    token_response = session.generate_token()

    return token_response


In [25]:


token_response = generate_fyers_access_token(
    client_id=client_id,
    secret_key=secret_key,
    redirect_uri=redirect_uri
)

access_token = token_response["access_token"]
print("Access Token:", access_token)


Access Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiZDoxIiwiZDoyIiwieDowIiwieDoxIiwieDoyIl0sImF0X2hhc2giOiJnQUFBQUFCcDFMd3daYkp5UlRpVU9uRjdxX3NqX0llNmU0Rm01YVpOQWFOT2ZyOWRrVnlRb2xJQlhLc0tzNGk2YzRrWG5HOHhpY2VhRlRSSWQwT0ZDT1BvUWFuWXN0cHpOMTVQb3hjeXY0NzdqNDU5Uzg2aGtCaz0iLCJkaXNwbGF5X25hbWUiOiIiLCJvbXMiOiJLMSIsImhzbV9rZXkiOiIwZjk0MWIyMzk2NTZkZjdjNGQ3MzZmOTk4NTE2M2ZjZjQxODRiOWJiOTk1MzAxNWI4ZWNiOTdlMCIsImlzRGRwaUVuYWJsZWQiOiJOIiwiaXNNdGZFbmFibGVkIjoiTiIsImZ5X2lkIjoiWEwwMDcyMSIsImFwcFR5cGUiOjIwMCwiZXhwIjoxNzc1NjA4MjAwLCJpYXQiOjE3NzU1NDk0ODgsImlzcyI6ImFwaS5meWVycy5pbiIsIm5iZiI6MTc3NTU0OTQ4OCwic3ViIjoiYWNjZXNzX3Rva2VuIn0.F6l31aosuZ0kTjoIwWt4SYZkIYXcq18NBuMtHmPPNsM


In [26]:
client_id = client_id
# Initialize the FyersModel instance with your client_id, access_token, and enable async mode
fyers = fyersModel.FyersModel(client_id=client_id, token=access_token,is_async=False, log_path="")

data = {
    "symbol":"NSE:IDEA-EQ",
    "qty":1,
    "type":1,
    "side":1,
    "productType":"INTRADAY",
    "limitPrice":0,
    "stopPrice":0,
    "validity":"DAY",
    "disclosedQty":0,
    "offlineOrder":False,
    "orderTag":"tag1",
    "isSliceOrder":False
}
response = fyers.place_order(data=data)
print(response)


{'code': -50, 'message': 'Request rejected: Orders are only allowed from whitelisted IP addresses. This request was received from IP: 2401:4900:1c5e:dd67:3806:edc2:61ae:d2a5.', 's': 'error'}


In [27]:
access_token

'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiZDoxIiwiZDoyIiwieDowIiwieDoxIiwieDoyIl0sImF0X2hhc2giOiJnQUFBQUFCcDFMd3daYkp5UlRpVU9uRjdxX3NqX0llNmU0Rm01YVpOQWFOT2ZyOWRrVnlRb2xJQlhLc0tzNGk2YzRrWG5HOHhpY2VhRlRSSWQwT0ZDT1BvUWFuWXN0cHpOMTVQb3hjeXY0NzdqNDU5Uzg2aGtCaz0iLCJkaXNwbGF5X25hbWUiOiIiLCJvbXMiOiJLMSIsImhzbV9rZXkiOiIwZjk0MWIyMzk2NTZkZjdjNGQ3MzZmOTk4NTE2M2ZjZjQxODRiOWJiOTk1MzAxNWI4ZWNiOTdlMCIsImlzRGRwaUVuYWJsZWQiOiJOIiwiaXNNdGZFbmFibGVkIjoiTiIsImZ5X2lkIjoiWEwwMDcyMSIsImFwcFR5cGUiOjIwMCwiZXhwIjoxNzc1NjA4MjAwLCJpYXQiOjE3NzU1NDk0ODgsImlzcyI6ImFwaS5meWVycy5pbiIsIm5iZiI6MTc3NTU0OTQ4OCwic3ViIjoiYWNjZXNzX3Rva2VuIn0.F6l31aosuZ0kTjoIwWt4SYZkIYXcq18NBuMtHmPPNsM'

    Tradebook
- Fetches all trades for the current day

In [28]:
from fyers_apiv3 import fyersModel

client_id = client_id
access_token = access_token
# Initialize the FyersModel instance with your client_id, access_token, and enable async mode
fyers = fyersModel.FyersModel(client_id=client_id, token=access_token,is_async=False, log_path="")

response = fyers.tradebook()
print(response)

{'code': 200, 'message': '', 's': 'ok', 'tradeBook': [{'clientId': 'XL00721', 'exchange': 10, 'fyToken': '10100000003456', 'orderNumber': '26040700417177', 'exchangeOrderNo': '1300000062139707', 'tradeNumber': '26040700417177-605419421', 'tradePrice': 305.75, 'segment': 10, 'productType': 'INTRADAY', 'tradedQty': 5, 'symbol': 'NSE:TMPV-EQ', 'row': 1775548171, 'orderDateTime': '07-Apr-2026 13:19:31', 'tradeValue': 1528.75, 'side': 1, 'orderType': 1, 'orderTag': '1:donchianBreakout'}, {'clientId': 'XL00721', 'exchange': 10, 'fyToken': '10100000003456', 'orderNumber': '26040700278748', 'exchangeOrderNo': '1300000041163100', 'tradeNumber': '26040700278748-603791802', 'tradePrice': 304.5, 'segment': 10, 'productType': 'INTRADAY', 'tradedQty': 5, 'symbol': 'NSE:TMPV-EQ', 'row': 1775541846, 'orderDateTime': '07-Apr-2026 11:34:06', 'tradeValue': 1522.5, 'side': -1, 'orderType': 2, 'orderTag': '1:Positions'}, {'clientId': 'XL00721', 'exchange': 10, 'fyToken': '10100000003456', 'orderNumber': '2

In [29]:
response["tradeBook"]

[{'clientId': 'XL00721',
  'exchange': 10,
  'fyToken': '10100000003456',
  'orderNumber': '26040700417177',
  'exchangeOrderNo': '1300000062139707',
  'tradeNumber': '26040700417177-605419421',
  'tradePrice': 305.75,
  'segment': 10,
  'productType': 'INTRADAY',
  'tradedQty': 5,
  'symbol': 'NSE:TMPV-EQ',
  'row': 1775548171,
  'orderDateTime': '07-Apr-2026 13:19:31',
  'tradeValue': 1528.75,
  'side': 1,
  'orderType': 1,
  'orderTag': '1:donchianBreakout'},
 {'clientId': 'XL00721',
  'exchange': 10,
  'fyToken': '10100000003456',
  'orderNumber': '26040700278748',
  'exchangeOrderNo': '1300000041163100',
  'tradeNumber': '26040700278748-603791802',
  'tradePrice': 304.5,
  'segment': 10,
  'productType': 'INTRADAY',
  'tradedQty': 5,
  'symbol': 'NSE:TMPV-EQ',
  'row': 1775541846,
  'orderDateTime': '07-Apr-2026 11:34:06',
  'tradeValue': 1522.5,
  'side': -1,
  'orderType': 2,
  'orderTag': '1:Positions'},
 {'clientId': 'XL00721',
  'exchange': 10,
  'fyToken': '10100000003456',


In [30]:
# Initialize the FyersModel instance with your client_id, access_token, and enable async mode
fyers = fyersModel.FyersModel(client_id=client_id, token=access_token,is_async=False, log_path="")

response = fyers.orderbook()
print(response)

{'code': 200, 'message': '', 's': 'ok', 'orderBook': [{'clientId': 'XL00721', 'exchange': 10, 'fyToken': '10100000003456', 'id': '26040700278464', 'id_fyers': 'ab595230-4f43-4e5b-b7c6-f06ad8099e1f', 'instrument': 0, 'offlineOrder': False, 'source': 'API', 'status': 2, 'type': 1, 'pan': 'DGEPR9605D', 'limitPrice': 309.05, 'productType': 'INTRADAY', 'qty': 5, 'disclosedQty': 0, 'remainingQuantity': 0, 'segment': 10, 'symbol': 'NSE:TMPV-EQ', 'description': 'TATA MOTORS PASS VEH LTD', 'ex_sym': 'TMPV', 'orderDateTime': '07-Apr-2026 11:33:42', 'side': 1, 'orderValidity': 'DAY', 'stopPrice': 0, 'tradedPrice': 304.6, 'filledQty': 5, 'exchOrdId': '1300000041105049', 'message': '', 'ch': -0.65, 'chp': -0.2114508783344177, 'lp': 306.75, 'orderNumStatus': '26040700278464:2', 'slNo': 1, 'orderTag': '1:donchianBreakout', 'takeProfit': 0, 'stopLoss': 0}, {'clientId': 'XL00721', 'exchange': 10, 'fyToken': '10100000003456', 'id': '26040700278748', 'id_fyers': 'ee801c07-3b63-44d8-8771-fd7574923dc7ADR',

    Orders

- For current day

In [31]:
response["orderBook"]

[{'clientId': 'XL00721',
  'exchange': 10,
  'fyToken': '10100000003456',
  'id': '26040700278464',
  'id_fyers': 'ab595230-4f43-4e5b-b7c6-f06ad8099e1f',
  'instrument': 0,
  'offlineOrder': False,
  'source': 'API',
  'status': 2,
  'type': 1,
  'pan': 'DGEPR9605D',
  'limitPrice': 309.05,
  'productType': 'INTRADAY',
  'qty': 5,
  'disclosedQty': 0,
  'remainingQuantity': 0,
  'segment': 10,
  'symbol': 'NSE:TMPV-EQ',
  'description': 'TATA MOTORS PASS VEH LTD',
  'ex_sym': 'TMPV',
  'orderDateTime': '07-Apr-2026 11:33:42',
  'side': 1,
  'orderValidity': 'DAY',
  'stopPrice': 0,
  'tradedPrice': 304.6,
  'filledQty': 5,
  'exchOrdId': '1300000041105049',
  'message': '',
  'ch': -0.65,
  'chp': -0.2114508783344177,
  'lp': 306.75,
  'orderNumStatus': '26040700278464:2',
  'slNo': 1,
  'orderTag': '1:donchianBreakout',
  'takeProfit': 0,
  'stopLoss': 0},
 {'clientId': 'XL00721',
  'exchange': 10,
  'fyToken': '10100000003456',
  'id': '26040700278748',
  'id_fyers': 'ee801c07-3b63-4

    Positions
-   Fetches current day's open and closed positions

In [32]:
# Initialize the FyersModel instance with your client_id, access_token, and enable async mode
fyers = fyersModel.FyersModel(client_id=client_id, token=access_token,is_async=False, log_path="")

response = fyers.positions()
print(response)

{'code': 200, 'message': '', 's': 'ok', 'netPositions': [{'symbol': 'NSE:TMPV-EQ', 'id': 'NSE:TMPV-EQ-INTRADAY', 'buyAvg': 305.175, 'buyQty': 10, 'buyVal': 3051.75, 'sellAvg': 305.025, 'sellQty': 10, 'sellVal': 3050.25, 'netAvg': 0, 'netQty': 0, 'side': 0, 'qty': 0, 'productType': 'INTRADAY', 'realized_profit': -1.500000000000341, 'crossCurrency': '', 'rbiRefRate': 1, 'fyToken': '10100000003456', 'exchange': 10, 'segment': 10, 'dayBuyQty': 10, 'daySellQty': 10, 'cfBuyQty': 0, 'cfSellQty': 0, 'qtyMulti_com': 1, 'pl': -1.500000000000341, 'unrealized_profit': 0, 'ltp': 306.75, 'slNo': 0}], 'overall': {'count_open': 0, 'count_total': 1, 'pl_realized': -1.500000000000341, 'pl_total': -1.500000000000341, 'pl_unrealized': 0}}


In [33]:
response

{'code': 200,
 'message': '',
 's': 'ok',
 'netPositions': [{'symbol': 'NSE:TMPV-EQ',
   'id': 'NSE:TMPV-EQ-INTRADAY',
   'buyAvg': 305.175,
   'buyQty': 10,
   'buyVal': 3051.75,
   'sellAvg': 305.025,
   'sellQty': 10,
   'sellVal': 3050.25,
   'netAvg': 0,
   'netQty': 0,
   'side': 0,
   'qty': 0,
   'productType': 'INTRADAY',
   'realized_profit': -1.500000000000341,
   'crossCurrency': '',
   'rbiRefRate': 1,
   'fyToken': '10100000003456',
   'exchange': 10,
   'segment': 10,
   'dayBuyQty': 10,
   'daySellQty': 10,
   'cfBuyQty': 0,
   'cfSellQty': 0,
   'qtyMulti_com': 1,
   'pl': -1.500000000000341,
   'unrealized_profit': 0,
   'ltp': 306.75,
   'slNo': 0}],
 'overall': {'count_open': 0,
  'count_total': 1,
  'pl_realized': -1.500000000000341,
  'pl_total': -1.500000000000341,
  'pl_unrealized': 0}}

In [34]:
response["overall"]

{'count_open': 0,
 'count_total': 1,
 'pl_realized': -1.500000000000341,
 'pl_total': -1.500000000000341,
 'pl_unrealized': 0}

In [35]:
response["netPositions"]

[{'symbol': 'NSE:TMPV-EQ',
  'id': 'NSE:TMPV-EQ-INTRADAY',
  'buyAvg': 305.175,
  'buyQty': 10,
  'buyVal': 3051.75,
  'sellAvg': 305.025,
  'sellQty': 10,
  'sellVal': 3050.25,
  'netAvg': 0,
  'netQty': 0,
  'side': 0,
  'qty': 0,
  'productType': 'INTRADAY',
  'realized_profit': -1.500000000000341,
  'crossCurrency': '',
  'rbiRefRate': 1,
  'fyToken': '10100000003456',
  'exchange': 10,
  'segment': 10,
  'dayBuyQty': 10,
  'daySellQty': 10,
  'cfBuyQty': 0,
  'cfSellQty': 0,
  'qtyMulti_com': 1,
  'pl': -1.500000000000341,
  'unrealized_profit': 0,
  'ltp': 306.75,
  'slNo': 0}]

In [38]:
data = {
    "symbols":"NSE:TMPV-EQ"
}
response = fyers.quotes(data=data)

{'code': 200,
 'message': '',
 's': 'ok',
 'netPositions': [{'symbol': 'NSE:TMPV-EQ',
   'id': 'NSE:TMPV-EQ-INTRADAY',
   'buyAvg': 305.175,
   'buyQty': 10,
   'buyVal': 3051.75,
   'sellAvg': 304.5,
   'sellQty': 5,
   'sellVal': 1522.5,
   'netAvg': 305.175,
   'netQty': 5,
   'side': 1,
   'qty': 5,
   'productType': 'INTRADAY',
   'realized_profit': -3.375000000000057,
   'crossCurrency': '',
   'rbiRefRate': 1,
   'fyToken': '10100000003456',
   'exchange': 10,
   'segment': 10,
   'dayBuyQty': 10,
   'daySellQty': 5,
   'cfBuyQty': 0,
   'cfSellQty': 0,
   'qtyMulti_com': 1,
   'pl': -0.7500000000001705,
   'unrealized_profit': 2.6249999999998863,
   'ltp': 305.7,
   'slNo': 0}],
 'overall': {'count_open': 1,
  'count_total': 1,
  'pl_realized': -3.375000000000057,
  'pl_total': -0.7500000000001705,
  'pl_unrealized': 2.6249999999998863}}

[{'symbol': 'NSE:TMPV-EQ',
  'id': 'NSE:TMPV-EQ-INTRADAY',
  'buyAvg': 304.6,
  'buyQty': 5,
  'buyVal': 1523,
  'sellAvg': 304.5,
  'sellQty': 5,
  'sellVal': 1522.5,
  'netAvg': 0,
  'netQty': 0,
  'side': 0,
  'qty': 0,
  'productType': 'INTRADAY',
  'realized_profit': -0.5000000000001137,
  'crossCurrency': '',
  'rbiRefRate': 1,
  'fyToken': '10100000003456',
  'exchange': 10,
  'segment': 10,
  'dayBuyQty': 5,
  'daySellQty': 5,
  'cfBuyQty': 0,
  'cfSellQty': 0,
  'qtyMulti_com': 1,
  'pl': -0.5000000000001137,
  'unrealized_profit': 0,
  'ltp': 305.7,
  'slNo': 0}]

{'count_open': 0,
 'count_total': 1,
 'pl_realized': -1.500000000000341,
 'pl_total': -1.500000000000341,
 'pl_unrealized': 0}

[{'symbol': 'NSE:TMPV-EQ',
  'id': 'NSE:TMPV-EQ-INTRADAY',
  'buyAvg': 305.175,
  'buyQty': 10,
  'buyVal': 3051.75,
  'sellAvg': 305.025,
  'sellQty': 10,
  'sellVal': 3050.25,
  'netAvg': 0,
  'netQty': 0,
  'side': 0,
  'qty': 0,
  'productType': 'INTRADAY',
  'realized_profit': -1.500000000000341,
  'crossCurrency': '',
  'rbiRefRate': 1,
  'fyToken': '10100000003456',
  'exchange': 10,
  'segment': 10,
  'dayBuyQty': 10,
  'daySellQty': 10,
  'cfBuyQty': 0,
  'cfSellQty': 0,
  'qtyMulti_com': 1,
  'pl': -1.500000000000341,
  'unrealized_profit': 0,
  'ltp': 305.55,
  'slNo': 0}]

In [ ]:
response

{'message': '',
 'code': 200,
 'd': [{'n': 'NSE:TMPV-EQ',
   'v': {'ask': 307,
    'bid': 306.95,
    'chp': -0.13,
    'ch': -0.4,
    'description': 'NSE:TMPV-EQ',
    'exchange': 'NSE',
    'fyToken': '10100000003456',
    'high_price': 307,
    'low_price': 301.85,
    'lp': 307,
    'open_price': 306.5,
    'original_name': 'NSE:TMPV-EQ',
    'prev_close_price': 307.4,
    'short_name': 'TMPV-EQ',
    'spread': 0.05,
    'symbol': 'NSE:TMPV-EQ',
    'tt': '1775520000',
    'volume': 4815616,
    'atp': 304.5},
   's': 'ok'}],
 's': 'ok'}

: 